builder/dockerfile/internals.go

```diff
@@ -38,41 +38,42 @@ import (
 	"github.com/pkg/errors"
 )
 
-func (b *Builder) commit(id string, autoCmd strslice.StrSlice, comment string) error {
+func (b *Builder) commit(comment string) error {
 	if b.disableCommit {
 		return nil
 	}
 	if !b.hasFromImage() {
 		return errors.New("Please provide a source image with `from` prior to commit")
 	}
+	// TODO: why is this set here?
 	b.runConfig.Image = b.image
 
-	if id == "" {
-		cmd := b.runConfig.Cmd
-		b.runConfig.Cmd = strslice.StrSlice(append(getShell(b.runConfig), "#(nop) ", comment))
-		defer func(cmd strslice.StrSlice) { b.runConfig.Cmd = cmd }(cmd)
-
-		hit, err := b.probeCache()
-		if err != nil {
-			return err
-		} else if hit {
-			return nil
-		}
-		id, err = b.create()
-		if err != nil {
-			return err
-		}
+	runConfigWithCommentCmd := copyRunConfig(b.runConfig, withCmdComment(comment))
+	hit, err := b.probeCache(b.image, runConfigWithCommentCmd)
+	switch {
+	case err != nil:
+		return err
+	case hit:
+		return nil
+	}
+	id, err := b.create(runConfigWithCommentCmd)
+	if err != nil {
+		return err
 	}
 
-	// Note: Actually copy the struct
-	autoConfig := *b.runConfig
-	autoConfig.Cmd = autoCmd
+	return b.commitWithID(id, b.runConfig)
+}
+
+func (b *Builder) commitWithID(id string, runConfig *container.Config) error {
```

Comment:
I think in this case it doesn't really work. These functions could actually probably be renamed.  `commitWithID()` should actually be `commit()`, and what is currently `commit()` should be renamed to `probeAndCommit()`